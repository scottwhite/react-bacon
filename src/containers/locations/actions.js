export const locationActions = {
  locationChange(evnt){
    if (evnt.charCode === 13){
      let search = evnt.target.value;
      return{
        type: 'LOCATIONS_GET_REQUESTED',
        search
      }
    }
    return { type: 'NOTHING' }
  },
  locationDetails(placeid){
    return {
      type: 'LOCATION_DETAILS_GET_REQUESTED',
      placeid
    }
  },
  weatherFor(latlng){
    let path = '/weather/' + latlng
    return {
      type: 'ROUTE_TO',
      path
    }
  }
}

export default locationActions
