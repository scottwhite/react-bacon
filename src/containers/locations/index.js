import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Places from './places';
import locationActions from './actions'

export class Location extends Component {
  render() {
    const {error, locations, locationDetails, locationChange} = this.props;
    let err;
    if(error){
      err = (<div className='alert alert-danger'>{error}</div>)
    }
    return (
      <div className='container-fluid'>
        {err}
        <div>Search for a place for weather</div>
        <input onKeyPress={locationChange} />
        <Places locations={locations || []} onSelected={locationDetails} />
      </div>
    );
  }
}

const mapStateToProps = (state) => {
  const { locations, error } = state.locations;
  return {
    locations,
    error
  };
};
const mapDispatchToProps = (dispatch) => {
  return bindActionCreators(locationActions, dispatch);
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(Location);
