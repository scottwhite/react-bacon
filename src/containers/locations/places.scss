.places-list{
  list-style: none;
  li {
    cursor: pointer;
  }
  a{
    color: blue !important;
  }
}
