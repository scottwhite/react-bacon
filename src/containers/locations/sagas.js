import { call, put } from 'redux-saga/effects'
import { push as routerpush } from 'react-router-redux'
import Api from '../../api'

// worker Saga: will be fired on LOCATIONS_GET_REQUESTED actions
export function* fetchLocations(action) {
  try {
    const data = yield call(Api.locations, action.search)
    yield put({type: 'LOCATIONS_GET_SUCCEEDED', locations: data})
  } catch (e) {
    yield put({type: 'LOCATIONS_GET_FAILED', message: e.message})
  }
}
// worker Saga: will be fired on LOCATION_DETAILS_GET_REQUESTED actions
export function* fetchLocationDetails(action) {
  function latLng(data){
    let {lat, lng} = data.result.geometry.location
    return lat + ', ' + lng
  }
  try {
    const data = yield call(Api.location_details, action.placeid)
    yield put({type: 'LOCATION_DETAILS_GET_SUCCEEDED', details: data})
    yield put(routerpush('/weather/' + latLng(data)))
  } catch (e) {
    console.error(e)
    yield put({type: 'LOCATION_DETAILS_GET_FAILED', message: e.message})
  }
}
