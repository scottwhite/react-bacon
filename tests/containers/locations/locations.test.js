import React from 'react';
import renderer from 'react/lib/ReactTestRenderer';

import {Location} from '~/src/containers/locations/index';
import locationActions from '~/src/containers/locations/actions'
import {places} from '~/mock-data/api.locations.mock'

describe('Location', () => {
  it('has an input that searches', () => {
    let locationChangeReal = locationActions.locationChange
    let search = 'washing'
    let locationChangeMock = (evnt)=>{
      let message = locationChangeReal(evnt)
      if(evnt.charCode === 13){
        expect(message.type).toBe('LOCATIONS_GET_REQUESTED')
        expect(message.search).toBe(search)
      }
    }

    const component = renderer.create(
      <Location locationChange={locationChangeMock}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot()

    let evnt = {charCode: 13, target: {value: search}}
    tree.children[1].props.onKeyPress(evnt)
  });

  it('shows a list if places that you can select an item', ()=>{
    let locationDetailsReal = locationActions.locationDetails
    let target = places.predictions[1]
    let targetid = target.place_id
    let locationDetailsMock = (placeid)=>{
      let message = locationDetailsReal(placeid)
      expect(message.type).toBe('LOCATION_DETAILS_GET_REQUESTED')
      expect(placeid).toBe(targetid)
    }
    const component = renderer.create(
      <Location locations={places.predictions} locationDetails={locationDetailsMock}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot()
    // BRITTLE!!!!!!
    // proposal for built in selectors
    // https://github.com/facebook/react/pull/7409
    tree.children[2].children[1].children[1].children[0].props.onClick()
  })

  it('shows an error if error', ()=>{
    let error = 'ERROR!'
    const component = renderer.create(
      <Location error={error}/>
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot()
  })

});
