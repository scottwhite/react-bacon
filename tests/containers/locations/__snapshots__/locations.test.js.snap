exports[`Location has an input that searches 1`] = `
<div
  className="container-fluid">
  <div>
    Search for a place for weather
  </div>
  <input
    onKeyPress={[Function locationChangeMock]} />
  <div
    className="container-fluid">
    <h2>
      Places
    </h2>
    <ul
      className="places-list" />
  </div>
</div>
`;

exports[`Location shows a list if places that you can select an item 1`] = `
<div
  className="container-fluid">
  <div>
    Search for a place for weather
  </div>
  <input
    onKeyPress={undefined} />
  <div
    className="container-fluid">
    <h2>
      Places
    </h2>
    <ul
      className="places-list">
      <li>
        <a
          onClick={[Function bound selectPlace]}>
          Washington, DC, United States
        </a>
      </li>
      <li>
        <a
          onClick={[Function bound selectPlace]}>
          Washington Boulevard, Jessup, MD, United States
        </a>
      </li>
      <li>
        <a
          onClick={[Function bound selectPlace]}>
          Washington Monument, Washington, DC, United States
        </a>
      </li>
      <li>
        <a
          onClick={[Function bound selectPlace]}>
          Washington Road, Edgewater, MD, United States
        </a>
      </li>
      <li>
        <a
          onClick={[Function bound selectPlace]}>
          Washington Boulevard, Laurel, MD, United States
        </a>
      </li>
    </ul>
  </div>
</div>
`;

exports[`Location shows an error if error 1`] = `
<div
  className="container-fluid">
  <div
    className="alert alert-danger">
    ERROR!
  </div>
  <div>
    Search for a place for weather
  </div>
  <input
    onKeyPress={undefined} />
  <div
    className="container-fluid">
    <h2>
      Places
    </h2>
    <ul
      className="places-list" />
  </div>
</div>
`;
